### PR TITLE
Minor bug fix to expectation_from_samples

### LINKED ISSUE #1535: Bug in `SymbolicHamiltonian.expectation_from_samples` for Qibo 0.2.14
**To Reproduce**
```
from qibo import Circuit, gates
from qibo.hamiltonians import SymbolicHamiltonian
from qibo.symbols import Z

circuit = Circuit(3)
circuit.add(gates.M(_i) for _i in (0, 2))
result = circuit(nshots=10)
freq = result.frequencies() # Counter({'00': 10})

ham = SymbolicHamiltonian(Z(2))
qubit_map = [0, 2]

print(ham.expectation_from_samples(freq, qubit_map))
```

The above code should simply give 1.0, but fails in Qibo 0.2.14. The error message:

```
Traceback (most recent call last):
  File "/home/wongzc/testing/sample_expectation/qibo_0p2p14_bug/bug.py", line 19, in <module>
    print(ham.expectation_from_samples(freq, qubit_map))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wongzc/miniforge3/envs/qibochem/lib/python3.11/site-packages/qibo/hamiltonians/hamiltonians.py", line 662, in expectation_from_samples
    [
  File "/home/wongzc/miniforge3/envs/qibochem/lib/python3.11/site-packages/qibo/hamiltonians/hamiltonians.py", line 664, in <listcomp>
    * (-1) ** [state[qubit_map[q]] for q in qubits].count("1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wongzc/miniforge3/envs/qibochem/lib/python3.11/site-packages/qibo/hamiltonians/hamiltonians.py", line 664, in <listcomp>
    * (-1) ** [state[qubit_map[q]] for q in qubits].count("1")
                     ~~~~~~~~~^^^
IndexError: list index out of range
```

I believe the problem is caused by the `qubit_map` argument in `SymbolicHamiltonian.expectation_from_samples`.

The documentation says that it should be a dictionary:
https://github.com/qiboteam/qibo/blob/a518e99fafd965b1a2d4271e2c095dd9574d0473/src/qibo/hamiltonians/hamiltonians.py#L638

But the default value is still a list:
https://github.com/qiboteam/qibo/blob/a518e99fafd965b1a2d4271e2c095dd9574d0473/src/qibo/hamiltonians/hamiltonians.py#L651-L652

while the code that checks the binary strings from the measurement frequencies will work if `qubit_map` is a dictionary (like `{0: 0, 2:1}` for my example above)
https://github.com/qiboteam/qibo/blob/a518e99fafd965b1a2d4271e2c095dd9574d0473/src/qibo/hamiltonians/hamiltonians.py#L664

**Suggested fix**

I would suggest keeping the `qubit_map` argument for `SymbolicHamiltonian` as a list (just edit the docs) to keep it consistent across all the different Hamiltonians available, then just change the line above to

```* (-1) ** [state[qubit_map.index(q)] for q in qubits].count("1")```

### PR DESCRIPTION
**Pull request by chmwzc**
Resolves #1535. 

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by chmwzc** [comment]
Hmm, the `expectation_from_circuit` function is affected by my change, let me look at that too...

**Comment by BrunoLiegiBastonLiegi** [comment]
Yep. most likely you have to update this 

https://github.com/qiboteam/qibo/blob/022bc925fb98246494d424f0d8255bf8121079d2/src/qibo/hamiltonians/hamiltonians.py#L613

in the same way.

**Comment by chmwzc** [comment]
- Took out the `qubit_map` argument from the `expectation_from_circuit` function and obtained it on the fly for each term in the Hamiltonian instead. E.g. for `Z(0)*Z(1) + Z(2)` `qubit_map` is set to be `[0, 1]` and `[2,]` respectively instead of using `[0, 1, 2]` for both terms.
- Also fixed some problems with `term.target_qubits` when the Hamiltonian has `I` terms (I use them sometimes to explain stuff to students)

File: src/qibo/hamiltonians/hamiltonians.py [hunk 1]
```diff
@@ -572,70 +570,70 @@ def expectation_from_circuit(
...
             nshots (int): number of shots, defaults to 1000.
 
         Returns:
             (float): the calculated expectation value.
         """
-        from qibo import Circuit, gates
-
-        if qubit_map is None:
-            qubit_map = list(range(self.nqubits))
+        from qibo import gates
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/hamiltonians/hamiltonians.py:578]: ah no sorry it was there already, I guess it was causing a circular import...

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]

**Review by Edoardo-Pedicillo — APPROVED** [review]

**Pull request merged**